### ISSUE TITLE
How to take the log of a negative number

### ISSUE THREAD

**Issue by simonzihlmann** [pyqtgraph/pyqtgraph#1988]
I am opening this issue to ask about how to take the log of a negative number as I already mentioned here: https://github.com/pyqtgraph/pyqtgraph/pull/1940

I personally find it a bit strange how it is currently done. If I understand it correctly, the log is taken like this: `y = np.sign(y) * np.log10(np.abs(y)+eps)`  with `eps = np.finfo(y.dtype).eps` or `eps = 1` if y is not a float.

For me this does not make sense. The logarithm of a negative number is not defined first of all. Then to go around that problem, I kind of agree to take the absolute value of the number and then to take the log. Because this tells you how "big" a number is. Multiplying with the sign after taking the log does not make sense at all since there is a lot of information in the sign of the output of the log. It tells you whether the number was bigger or smaller than the base 10.

Let me do a small example.
y = 1e2
--> in log mode: y = 1*(2+eps) ~2 (neglecting eps)

y = -1e-5
--> in log modde: y = -1*(-5+eps)~5 (neglecting eps)

Therefore, -1e5 appears to be 3 orders of magnitude bigger than 1e2 even though the absolute value is 7 orders of magnitude smaller. Therefore, y = np.sign(y) * np.log10(np.abs(y)+eps) leads to very strange plots that are really misleading if very small negative numbers are involved.

I would propose to use y = np.log10(np.abs(y)+eps)

I am also questioning the use of eps. This could also lead to strange behaviors. Why not setting the log of y to np.nan if it is zero? Strictly speaking, log(0)-->-inf, which cannot be plotted.

This is my opinion on this and I came across this since I had some strange plots when using log for the y-scale. Maybe there is a good reasoning behind how you do. I would be curious to know. Thanks anyway for the great library!

**Comment by j9ac9k** [comment]
I'll first say I'm open to a different implementation, the implementation that #1940 re-implements was created in #1476 so that may be of interest regarding reference.  If we do change up behavior here, I think we should get more input from other folks; so going to tag @NilsNemitz @ixjlyons @outofculture @ksunden and @pijyoi

**Comment by NilsNemitz** [comment]
Hi @simonzihlmann !

I don't have much time to look at code these days, but I have a few quick comments on the log implementation:

The current method seems to implement a very specific symmetric log display. But if I understand correctly, this is supposed to have *three* regions: There's a negative region of -log10 (abs(value) ), used for large negative values. There's a positive region of +log10(abs(value) ), used for large positive values. Then I *think* there should be a linear region for values of small magnitude (e.g. -10 to +10). The linear region avoids the ambiguity that @simonzihlmann points out here.

I wouldn't use this kind of display in a scientific paper, but it might be useful when showing differences between "naturally sized" numbers, e.g. stocked items or votes: When there is a small difference, the plot would directly show the difference. When the difference is large, the plot shows order of magnitude without scaling to hide the small differences between other points.

If the current implementation really works as described above (with only two zones), then I would consider that problematic.

Overall I would prefer the same basic implementation that @simonzihlmann asks for (log of negative is NaN), but for my data, I usually know in advance if a plot needs log display, and just plot the log. Taking out the symmetric log option *might* be more painful for the users that rely on this than not having a proper log option is for me (and others?).

The ultimately best solution would be to support generic mapping functions (one of which would be log) rather than hardcoding various interpretations of log-scaling in multiple places. Maybe somebody has a good idea how to tackle that?

TLDR: 
- I can't tell if it's better to keep what we have or switch to a proper log display.
- The symmetric log as described above is indeed ambiguous and should be modified if kept.

**Comment by pijyoi** [comment]
Regarding use of ```DBL_EPSILON``` as _eps_, this is what happens.
```DBL_EPSILON``` is not an arbitrarily small number.
```python
import pyqtgraph as pg
pg.mkQApp()
pw = pg.PlotWidget()
pw.show()
pi = pw.getPlotItem()
pi.setLogMode(x=False, y=True)
x = range(-25, 5)
y = [(10.**p) for p in x]
pi.plot(x, y, symbol='o')
pg.exec()
```
![Screenshot 2021-09-10 143144](https://user-images.githubusercontent.com/2657027/132810350-3bc9909c-b4d2-4b71-bce0-86b5ae635238.png)

**Comment by NilsNemitz** [comment]
That looks like DBL_EPSILON is chosen to represent the resolution of the 52 bit mantissa of a double.
So it represents the *fractional* resolution of a double value. Numbers can be a lot smaller (< 1E-308), but a number of magnitude 1 will have a quantization error at ~2E-16.

I like to be able to plot 10^-18 numbers, too; so I would prefer not to add 2E-16 to everything.  :)

**Comment by simonzihlmann** [comment]
I agree, adding an arbitrary offset is not what I would expect from a plotting library. I am in favour of skipping DBL_EPSILON and to solve the zero value case with not a number (if this can be handled by pyqtgraph).

**Comment by j9ac9k** [comment]
We plot NaN values in a few different ways.  By default, we replace NaN values with the finite value _before_ it in the array.  With a performance hit, we could specify `connect='finite'`, take a performance hit (thanks to @pijyoi not a large one) and have the lines be discontinuous when we have NaN points (which is probably a better solution than using `eps`).

I'm somewhat curious how other libraries (matplotlib, bokeh and so on) handle log-scaling with negative values.

**Comment by j9ac9k** [comment]
Thinking back a little more, I'm going to agree w/ @simonzihlmann and say that merging #1476 was a mistake.  log does represent a 1 to 1 transformation, but since #1476 we no longer can make that guarantee (log(0.1) = log(-10) as we currently have it).

the epsilon bit is also troublesome for reasons posted above; should we convert 0 values to NaN and do the plot generation using `connect='finite'`?

**Comment by pijyoi** [comment]
A nitpick, log ( 0 ) returns -Inf. log of strictly negative values return NaN.

**Comment by j9ac9k** [comment]
> A nitpick, log ( 0 ) returns -Inf. log of strictly negative values return NaN.

after merging #1476 I think I deserve all the nit-picks on this subject for a while 😆

**Comment by pijyoi** [comment]
I didn't realize that ```np.sign(0) == 0```.
So ```np.sign(0) * np.log10(np.abs(0) + eps) == 0```

The intention of adding ```eps``` was probably to avoid evaluating ```log10(0)```, probably intending the final plotted value to be ```log10(eps)```.
```np.sign()``` probably defeated that intention.

```python
import numpy as np
import pyqtgraph as pg
pg.mkQApp()
pw = pg.PlotWidget()
pw.show()
pi = pw.getPlotItem()
pi.setLogMode(x=False, y=True)
x = range(-20, 0)
yp = np.array([(10.**p) for p in x])
y = np.hstack([0, yp])    # stuff a zero in front
pi.plot(y, symbol='o')
pg.exec()
```
![Capture](https://user-images.githubusercontent.com/2657027/132971167-ec3aa375-b2ea-48c7-ab5f-1d72e14fce6c.PNG)

**Comment by j9ac9k** [comment]
> I didn't realize that `np.sign(0) == 0`.

🤦🏻‍♂️  yeah I really screwed that one up; I'm all for reverting to previous behavior just take `np.log10(y)`; only question I have is can we plot inf/-inf; but it may take some work to get that to work in a reliable way, and perhaps that's a separate PR; and just let numpy emit its warning and the user take care of it.

**Comment by pijyoi** [comment]
> just take np.log10(y)

I tried that and it caused a Python interpreter crash if the user passed-in values are not strictly positive. A plot did get rendered, the crash occurs after.

**Comment by pijyoi** [comment]
If the code is changed to ```y = np.log10(y)```, then the following script hangs the GUI.
The hang requires all the following conditions to trigger:
1) change of code to ```y = np.log10(y)```
2) log mode enabled
3) inf in the data (nans are ok, however)
4) dynamicRangeLimit left enabled

```python
import numpy as np
import pyqtgraph as pg
app = pg.mkQApp()
pw = pg.PlotWidget()
pw.show()
pi = pw.getPlotItem()
pi.setLogMode(x=False, y=True)
x = range(-20, 20)
yp = np.array([(10.**p) for p in x])

# infs in the data require disabling dynamicRangeLimit by setting it to None
# default value of dynamicRangeLimit, if not set, is 1e6
pi.plot(np.hstack([0, yp]), symbol='o', dynamicRangeLimit=1e6) # log10(0) -> -inf

# nans in the data are fine
# pi.plot(np.hstack([-1, yp]), symbol='o')  # log10(-1) -> nan

app.exec()
```

**Comment by j9ac9k** [comment]
@NilsNemitz would you have time to evaluate what in the `dynamicRangeLimit` is the issue?  I can try and take a look but it may be a few days before I even start poking at it.

**Comment by NilsNemitz** [comment]
I just had a look, and the center of suspicion is currently
https://github.com/pyqtgraph/pyqtgraph/blob/master/pyqtgraph/graphicsItems/PlotDataItem.py#L766
The dynamic range limiter assumes that the limits available as `self.dataRect()` represents the data as it appears on screen. But this does not seem to be the case: `dataRect()` does not update to reflect the log coordinates.

The final breakdown seems to be a slow, endless spiral of redraw --> clipping adjustment --> redraw --> clipping adjustment.

The next step is probably to figure out if this can be fixed by adjusting `dataRect()` to account for log mode, or if special case handling in the dynamic range limiter is needed. I hope I will have some more time in the next few days.

I possible "solution" would be to disable the limiter in log mode. Log mode compresses the displayed data. I don't currently see a use case where it is necessary to zoom into log data to the extent that would trigger the Qt dynamic range issues.

**Comment by j9ac9k** [comment]
> A possible "solution" would be to simply disable the limiter in log mode. Log mode compresses the displayed data. I don't currently see a use case where it is necessary to zoom into log data to the extent that would trigger the Qt dynamic range issues.

I think in the near-term this is likely the best thing we should implement. 

I suspect the `dataRect()` specific issue is a consequence of just not having that non-linear transformation we can apply (which is a real downer!)

**Comment by pijyoi** [comment]
May be related. The following script doesn't hang the gui but seems to take several update cycles to stabilize / converge.
The final render has -inf and inf clipped to the dynamic range.
```python
from math import inf
import pyqtgraph as pg
pg.plot([1, 2, 3, -inf, 5, 6, inf, 8, 9, 10], symbol='o')
pg.exec()
```

There's also one runtime warning
```
c:\work\github\pyqtgraph\pyqtgraph\functions.py:1126: RuntimeWarning: invalid value encountered in multiply
  mapped = (m*coords).sum(axis=1)  ## apply scale/rotate
```

Probably infs are just not meant to be passed into PlotDataItem? I can imagine that the various Plot Options->Transforms aren't expecting non-finite values.

**Comment by NilsNemitz** [comment]
Well, PlotDataItem now certainly isn't set up to deal with infs. But in most cases, numpy handles those without too many complaints, and they do not seem to break the plots. So it might be possible to fix some bugs and allow infs to work. 

An alternative might be to replace all `inf` with `nan` ? But such a global search-and-replace seems somewhat inelegant. It's cheap compared to the log conversion, so we might add it only for log mode. But then it is still not very satisfying that any `inf` in a (non-log) dataset breaks stuff.

I am not sure why that warning triggers when `inf` is involved in the calculation, while `nan` is quietly accepted. I have been able to reproduce the warning in the python interpreter using the values of `m` and `coords` generated by the example code... but only for this kind of matrix operation, simple multiplication with `inf` and `nan` seems to give usable results without warnings. I thought it might be some spacial case like `inf * 0`, but that just seems to give `nan`, with no warning. In the interpreter, the warning is also emitted only once, but for some reason, the PyQtGraph use repeats it for any manipulation of the view. 

Despite the warning, the output looks healthy: Invalid input points generate invalid invalid coordinates, while good inputs make good coordinates. I think this is what we want to happen.

Since nothing actually breaks, and not plotting `inf` and `nan` is not really unexpected behavior, I would be in favor of suppressing the warning in line 1126 of functions.py:
```python 
    with warnings.catch_warnings():
        warnings.simplefilter("ignore", RuntimeWarning)
        mapped = (m*coords).sum(axis=1)  ## apply scale/rotate
```
Some more granular filtering might be better, but I didn't see a way to do that without spooling up a full regex engine...

**Comment by sem-geologist** [comment]
Well this will probably will sound very silly from my side, but for cases of data going down and below zero: Why we don't use some non-linear function which behaves much better, i.e. cubic root? symmetric around 0, highlights more variations near 0 while hiding differences of big numbers. No any workarounds are needed. Leave log scale for data which historically needs to plotted with log and is meant to have no zeroes or negative numbers. (Actually I am looking currently how to implement cubic root scale alongside log)

**Comment by NilsNemitz** [comment]
Well, log plots have some special properties that make them a must-have even today. The main ones are probably that 
- power laws show up as straight lines in a log-log plot, which makes it very convenient to understand unknown data.
- a lot of decay processes are exponential functions, which again clearly show as straight lines in a log-lin plot.

This use in identifying functional dependencies also makes it very useful to have the context menu toggle switch. It might tell you something about your dataset that you didn't know before.

It would be *very* nice though to have the option to plug in additional "mapping modules", like your cubic-root mapping. I would imagine these as a type of object that implements forward and backwards transforms that are then called in the appropriate places. Such an object might also handle caching the converted data separately from the raw data, and track properties like the minima/maxima now (poorly) handled by `dataRect()`.

Maybe we will learn more about the various places this needs to interact with while we track down some of the log mode bugs?
Hardcoding more transforms seems like it would lead to more headaches in the future. 😁 

If you have some good ideas, they would be very welcome!

**Comment by sem-geologist** [comment]
sqrt-sqrt would also show linear, as cbrt-cbrt would show linear for data which shows linear behaviour in log-log plot. sqrt has no problem with 0, but would need workaround for negative numbers, while cbrt has no problem at all. Numpy has cbrt function, thus it is not computing intense.

as for hardcoding: currently Log modes are hardcoded. For more types of non-linear axis probably whole universal mechanism for providing non-linear scaling should be developed and log mode moved from being hardcoded into one of such plug-in-able kind. But Probably I am one of first which are not much content with log, and looking to other possibilities. (actually for X-ray spectroscopy I would more like Square root type scaling for Y, but cube root would be more universal (for symetric to negative values) and also would scratch my itch)

**Comment by NilsNemitz** [comment]
Oh, I am 100% on your side. Adding more modular stuff is the way to go. And once somebody gets a good handle on this, modularizing the log code as one of multiple options would be great... The current hardcoded solution has bugs in various places. Look up there; we have a special issue tag for log-related bugs!

But there is a special magic to the log representation, although part of this is probably forever lost to a dimension populated by slide rules and those who know how to use them. 

Converting between bases is a simple multiplication (Was it log10(x) = ln(x) / ln(10) ?), so the choice of base isn't particularly magic if you just need to see if your data set decays exponentially (or double exponentially or some-such). If you got a lot of negative values in your data, that probably isn't the case, and trying it out probably won't teach you much. But having the toggle switch rather than a logarithm table means trying it out is free. I guess the library probably shouldn't blow up.

And the power law diagnostic can do a little more than just check for linear behavior. Here is a live display that's hiding in the back of my desktop right now:
![adev](https://user-images.githubusercontent.com/19742018/133425196-31dce8b4-67c4-41b5-9ad3-7b85123458e7.png)

The yellow curve shows good phase-noise limited behavior; The value falls as perfect 1/tau (exponent and slope = -1).
If there was a proper measurement running, it would probably be showing frequency noise, which appears as a straight line with a slope of -1/2, indicating a decay with 1/sqrt(tau). The greenish curves are trouble-makers that want to mess up the data. But since their slope is somewhat steeper than -1/2, their noise contribution will become small enough over time.

Of course there isn't much point in plotting *that* data in any way that isn't log-log. So this doesn't actually use any pyqtgraph functions... The logarithm is applied before the data goes in there.

If I see a slope of +1/3, I will be happy to switch to the cube-root display 😃

**Comment by edumur** [comment]
I just looked at how matplotlib was doing things.

First I ran this script, mimicking @pijyoi  :
```python
import numpy as np
import matplotlib.pyplot as plt

x = range(-20, 0)
yp = np.array([(10.**p) for p in x])

y = np.hstack([0, yp])    # stuff a zero in front

fig, ax = plt.subplots(1, 1)
ax.plot(y, 'o')

ax.set_yscale('log')
plt.show()
```

From which I got this:
![image](https://user-images.githubusercontent.com/13218882/133619429-1bd3a36b-fd2e-43e5-82df-a89e90c56152.png)

And matplotlib take the log as follow, from [here](https://github.com/matplotlib/matplotlib/blob/a4e81e7fb40327d9dc6712daddc8152c88aaf07e/lib/matplotlib/scale.py):
```python
class LogTransform(Transform):
    input_dims = output_dims = 1

    def __init__(self, base, nonpositive='clip'):
        super().__init__()
        if base <= 0 or base == 1:
            raise ValueError('The log base cannot be <= 0 or == 1')
        self.base = base
        self._clip = _api.check_getitem(
            {"clip": True, "mask": False}, nonpositive=nonpositive)

    def __str__(self):
        return "{}(base={}, nonpositive={!r})".format(
            type(self).__name__, self.base, "clip" if self._clip else "mask")

    def transform_non_affine(self, a):
        # Ignore invalid values due to nans being passed to the transform.
        with np.errstate(divide="ignore", invalid="ignore"):
            log = {np.e: np.log, 2: np.log2, 10: np.log10}.get(self.base)
            if log:  # If possible, do everything in a single call to NumPy.
                out = log(a)
            else:
                out = np.log(a)
                out /= np.log(self.base)
            if self._clip:
                # SVG spec says that conforming viewers must support values up
                # to 3.4e38 (C float); however experiments suggest that
                # Inkscape (which uses cairo for rendering) runs into cairo's
                # 24-bit limit (which is apparently shared by Agg).
                # Ghostscript (used for pdf rendering appears to overflow even
                # earlier, with the max value around 2 ** 15 for the tests to
                # pass. On the other hand, in practice, we want to clip beyond
                #     np.log10(np.nextafter(0, 1)) ~ -323
                # so 1000 seems safe.
                out[a <= 0] = -1000
        return out

    def inverted(self):
        return InvertedLogTransform(self.base)
```
